### ISSUE TITLE
weave does not set ip address in docker container in docker 1.7.0

### ISSUE THREAD

**Issue by dove-young** [weaveworks/weave#1233]
Run docker daemon on <code>docker0</code>, <code>10.10.10.1</code>

Launch weave network and run docker container via weave proxy.  

IP address was allocated in <code>weave status</code> was <code>10.128.0.2</code> but IP address in the docker container is actually <code>10.10.10.24</code> which was allocated by docker daemon.

This is docker daemon runs
````
[root@tivx94 ~]# ps -ef | grep docker
root      8558     1  0 Jul29 ?        00:00:32 /usr/bin/docker -d --log-level error -p /var/run/docker.pid -s overlay --graph=/var/lib/docker -H unix:///var/run/docker.sock 
````
weave launch
````
[root@tivx94 ~]# weave launch   1>&2 2> weave.launch.10.new.debug.log                                                                       
3b2676319c33b40d32484000cc7bd40fc68e706bed45de699c053eb20d18e7a7
[root@tivx94 ~]# docker logs weave                                                                                                          
weave 2015/07/29 16:18:37.847946 Command line options: map[iface:ethwe iprange:10.128.0.0/10 name:52:3f:88:88:08:ad nickname:tivx94 port:678
3 wait:20]
weave 2015/07/29 16:18:37.848039 Command line peers: []
weave 2015/07/29 16:18:37.848179 Communication between peers is unencrypted.
weave 2015/07/29 16:18:37.848233 Our name is 52:3f:88:88:08:ad(tivx94)
weave 2015/07/29 16:18:37.848238 Assuming quorum size of 1
INFO: 2015/07/29 16:18:37.859835 [docker] Using Docker API on unix:///var/run/docker.sock: &[ApiVersion=1.19 GitCommit=74e7a7a/1.7.0 GoVersi
on=go1.4.2 Os=linux Arch=amd64 KernelVersion=4.0.8-300.fc22.x86_64 PackageVersion=docker-1.7.0-6.git74e7a7a.fc22.x86_64 Version=1.7.0.fc22]
weave 2015/07/29 16:18:37.893707 Sniffing traffic on &{151 65535 ethwe da:1b:00:63:15:58 up|broadcast|multicast}
weave 2015/07/29 16:18:37.893803 Discovered our MAC da:1b:00:63:15:58
weave 2015/07/29 16:18:38.733563 Discovered local MAC 22:eb:81:b8:51:9e
[root@tivx94 ~]# 

````
launch dns and proxy

````
[root@tivx94 ~]# weave launch-dns  1>&2 2> weave.launch-dns.10.new.debug.log                                                                
53cbf53c9100b60a3147ad7972e17067feb0b435c9a77833e6f4c9487f8d06c3
[root@tivx94 ~]# 
[root@tivx94 ~]# weave launch-proxy --with-dns  1>&2 2> weave.launch-proxy.10.new.debug.log                                                 
f45d9a16641a32b2afb05ed6b48e5e2f88f86a6e320295e7db2f39780dc7fec0
[root@tivx94 ~]# 
````
then run a container and inspect it's IP address, it goes to 10.10.10.24
````
[root@tivx94 ~]# docker run -dti --hostname=pingme.weave.local --name pinghim fedora/php-netcat:4.2.8 nc -lk -p 4300                        
bd3adee6e3a6b4b63e83cc24bb1611c844b9fb9e5f04002d376e1a4682b402ac
[root@tivx94 ~]# docker inspect  --format='{{.NetworkSettings.IPAddress}}' pinghim                                                          
10.10.10.24
[root@tivx94 ~]#
````
check proxy log
````
[root@tivx94 ~]# docker logs weaveproxy 
INFO: 2015/07/29 16:19:06.914830 weave proxy 1.0.1
INFO: 2015/07/29 16:19:06.914949 Command line arguments: --with-dns
INFO: 2015/07/29 16:19:06.939518 proxy listening on 0.0.0.0:12375
INFO: 2015/07/29 16:19:35.462652 GET /v1.19/containers/weaveproxy/json
INFO: 2015/07/29 16:19:35.466638 GET /v1.19/containers/weaveproxy/logs?stderr=1&stdout=1&tail=all
INFO: 2015/07/29 16:19:40.218425 GET /v1.19/containers/weavedns/json
INFO: 2015/07/29 16:19:40.220925 GET /v1.19/containers/weavedns/logs?stderr=1&stdout=1&tail=all
INFO: 2015/07/29 16:19:48.935707 GET /v1.19/containers/weave/json
INFO: 2015/07/29 16:19:48.939627 GET /v1.19/containers/weave/logs?stderr=1&stdout=1&tail=all
INFO: 2015/07/29 16:19:59.677441 POST /v1.19/containers/create?name=pinghim
INFO: 2015/07/29 16:19:59.677944 Creating container with WEAVE_CIDR ""
INFO: 2015/07/29 16:19:59.715509 POST /v1.19/containers/bd3adee6e3a6b4b63e83cc24bb1611c844b9fb9e5f04002d376e1a4682b402ac/start
INFO: 2015/07/29 16:19:59.843745 Attaching container bd3adee6e3a6b4b63e83cc24bb1611c844b9fb9e5f04002d376e1a4682b402ac with WEAVE_CIDR "" to weave network
INFO: 2015/07/29 16:20:06.632100 GET /v1.19/containers/pinghim/json
INFO: 2015/07/29 16:20:10.528398 GET /v1.19/containers/weaveproxy/json
INFO: 2015/07/29 16:20:10.531026 GET /v1.19/containers/weaveproxy/logs?stderr=1&stdout=1&tail=all
[root@tivx94 ~]# 
````
check weave status, it would be 10.128.0.2
````
[root@tivx94 ~]# weave status 
weave router 1.0.1
Our name is 52:3f:88:88:08:ad(tivx94)
Encryption off
Peer discovery on
Sniffing traffic on &{151 65535 ethwe da:1b:00:63:15:58 up|broadcast|multicast}
MACs:
da:1b:00:63:15:58 -> 52:3f:88:88:08:ad(tivx94) (2015-07-29 16:18:37.893763195 +0000 UTC)
22:eb:81:b8:51:9e -> 52:3f:88:88:08:ad(tivx94) (2015-07-29 16:18:38.733473402 +0000 UTC)
da:13:97:e7:04:a1 -> 52:3f:88:88:08:ad(tivx94) (2015-07-29 16:18:57.961097234 +0000 UTC)
ee:2f:a6:56:48:9d -> 52:3f:88:88:08:ad(tivx94) (2015-07-29 16:19:59.993083042 +0000 UTC)
Peers:
52:3f:88:88:08:ad(tivx94) (v0) (UID 5054573935842922381)
Routes:
unicast:
52:3f:88:88:08:ad -> 00:00:00:00:00:00
broadcast:
52:3f:88:88:08:ad -> []
Direct Peers:
Reconnects:

Allocator range [10.128.0.0-10.192.0.0)
Owned Ranges:
  10.128.0.0 -> 52:3f:88:88:08:ad (tivx94) (v2)
Allocator default subnet: 10.128.0.0/10

weave DNS 1.0.1
Listen address :53
Fallback DNS config &{[9.0.148.50] [cn.ibm.com] 53 1 5 2}

Local domain weave.local.
Interface &{155 65535 ethwe da:13:97:e7:04:a1 up|broadcast|multicast}
Zone database:
bd3adee6e3a6: pingme.weave.local.[10.128.0.2]


weave proxy is running
[root@tivx94 ~]# 
````

Here is the docker version
```
[root@tivx94 ~]# docker version
Client version: 1.7.0.fc22
Client API version: 1.19
Package Version (client): docker-1.7.0-6.git74e7a7a.fc22.x86_64
Go version (client): go1.4.2
Git commit (client): 74e7a7a/1.7.0
OS/Arch (client): linux/amd64
Server version: 1.7.0.fc22
Server API version: 1.19
Package Version (server): docker-1.7.0-6.git74e7a7a.fc22.x86_64
Go version (server): go1.4.2
Git commit (server): 74e7a7a/1.7.0
OS/Arch (server): linux/amd64
[root@tivx94 ~]# 
```

**Comment by rade** [comment]
Application containers get both a docker network IP address and weave network IP address. The IP address reported by `{{.NetworkSettings.IPAddress}}` is the former.

**Comment by dove-young** [comment]
I've been confused.  <code>weavedns</code> also recorded the docker network IP.

```
[root@tivx94 ~]# docker exec pinghim ping -c 1 pingme.weave.local                                                                                                          
PING pingme.weave.local (10.10.10.24) 56(84) bytes of data.
64 bytes from pingme.weave.local (10.10.10.24): icmp_seq=1 ttl=64 time=0.055 ms

--- pingme.weave.local ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.055/0.055/0.055/0.000 ms
[root@tivx94 ~]# 
```

And also the container name was not recorded into <code>weavedns</code>

```
[root@tivx94 ~]# docker exec pinghim ping -c 1  pinghim
ping: unknown host pinghim
[root@tivx94 ~]# 
```

**Comment by rade** [comment]
> `weavedns` also recorded the docker network IP

That is impossible. The address must come from somewhere else.

I don't see a `eval $(weave proxy-env)` in any of your transcripts. Did you forget it?

**Comment by dove-young** [comment]
Yes I evaluated proxy-env. Here is today's try. 
```
Allocator range [10.128.0.0-10.192.0.0)
Owned Ranges:
  10.128.0.0 -> 52:3f:88:88:08:ad (tivx94) (v2)
Allocator default subnet: 10.128.0.0/10

weave DNS 1.0.1
Listen address :53
Fallback DNS config &{[9.0.148.50] [cn.ibm.com] 53 1 5 2}

Local domain weave.local.
Interface &{155 65535 ethwe da:13:97:e7:04:a1 up|broadcast|multicast}
Zone database:
bd3adee6e3a6: pingme.weave.local.[10.128.0.2]


weave proxy is running
[root@tivx94 ~]# eval $(weave proxy-env)                                                                                                                                   
[root@tivx94 ~]# env | grep -i docker
DOCKER_HOST=tcp://127.0.0.1:12375
[root@tivx94 ~]# 
[root@tivx94 ~]# 
[root@tivx94 ~]# docker run -dti --hostname=test.weave.local --name pingthem  fedora/php-netcat:4.2.8 nc -lk -p 4300
def9924c7e44e478f8e8b6997ce0f1f243dfba462e4809f58408a03cafed42f7
[root@tivx94 ~]# docker run -dti --hostname=test.weave.local --name pingtest  fedora/php-netcat:4.2.8 nc -lk -p 4300
bff31b3211c86db1305e8561db04587aa263eae7ab701313bad66bc5808cb0df
[root@tivx94 ~]# docker exec pingtest ping test.weave.local
PING test.weave.local (10.10.10.26) 56(84) bytes of data.
64 bytes from test.weave.local (10.10.10.26): icmp_seq=1 ttl=64 time=0.091 ms
64 bytes from test.weave.local (10.10.10.26): icmp_seq=2 ttl=64 time=0.107 ms
^C[root@tivx94 ~]# 
[root@tivx94 ~]# docker exec pingtest ping pingthem
ping: unknown host pingthem
[root@tivx94 ~]# 
```

And you can find <code>test.weave.local</code> was added to <code>weavedns</code> but container name <code>pingthem</code> was not. 

```
[root@tivx94 ~]# docker logs weavedns
INFO: 2015/07/29 16:18:57.865610 [main] WeaveDNS version 1.0.1
INFO: 2015/07/29 16:18:57.865658 [main] Waiting for mDNS interface ethwe to come up
INFO: 2015/07/29 16:18:57.966055 [main] Interface ethwe is up
INFO: 2015/07/29 16:18:57.966080 [main] Waiting for HTTP interface eth0 to come up
INFO: 2015/07/29 16:18:57.966420 [main] Interface eth0 is up
INFO: 2015/07/29 16:18:57.966630 [zonedb] Using mDNS on &{Index:155 MTU:65535 Name:ethwe HardwareAddr:da:13:97:e7:04:a1 Flags:up|broadcast|multicast}
INFO: 2015/07/29 16:18:57.977195 [docker] Using Docker API on unix:///var/run/docker.sock: &[Os=linux Arch=amd64 KernelVersion=4.0.8-300.fc22.x86_64 PackageVersion=docker-1.7.0-6.git74e7a7a.fc22.x86_64 Version=1.7.0.fc22 ApiVersion=1.19 GitCommit=74e7a7a/1.7.0 GoVersion=go1.4.2]
INFO: 2015/07/29 16:18:57.978279 [main] HTTP API listening on 10.10.10.23:6785
INFO: 2015/07/29 16:18:57.978361 [dns] Upstream server(s): &{Servers:[9.0.148.50] Search:[cn.ibm.com] Port:53 Ndots:1 Timeout:5 Attempts:2}
INFO: 2015/07/29 16:18:57.978375 [dns] Cache: 8192 entries
INFO: 2015/07/29 16:18:57.979690 [dns] Listening for DNS on :53 (UDP)
INFO: 2015/07/29 16:18:57.980175 [dns] Listening for DNS on :53 (TCP)
INFO: 2015/07/29 16:18:58.200023 [zonedb] Container df8641c2bb9414adaa7ba8b9f89514339e3e105488a376bb0c14228fb4b37ed6 down. Removing records
INFO: 2015/07/29 16:19:06.979394 [zonedb] Container 5dbc8f131686114725b73b4ae29a83a4657663a664aa56175662726fa124cfa0 down. Removing records
INFO: 2015/07/29 16:19:32.531755 [zonedb] Container 681b13fa3274b2edc41130b6e65e19123b9d5188de5d6201ceeeb03eed6f351f down. Removing records
INFO: 2015/07/29 16:20:00.088077 [http] Adding pingme.weave.local -> 10.128.0.2
INFO: 2015/07/29 16:25:16.117589 [zonedb] Container 6a81673e3081a64c9a8a08f0a67b2b53d94296da8172f0feab8133b7a4c625b6 down. Removing records
INFO: 2015/07/29 16:25:47.617459 [zonedb] Container c3306098f1e73ae7ae205a12dc90a537b0a4b9c2c4754173041be0de398a26f9 down. Removing records
INFO: 2015/07/29 16:25:58.084053 [zonedb] Container 0267a81a71cbb7e376c9fd460bb5d03e697e763538901e74a8d5a3eb2643ddaa down. Removing records
INFO: 2015/07/29 16:26:01.961548 [zonedb] Container 1e972a4cf49281eef05690bd184917219207d8c0a65594f595fb1c18d746e4d6 down. Removing records
INFO: 2015/07/29 16:49:46.967809 [zonedb] Container 46c214afb3c2f1e8116f2e3fe73f67879b50e09e9a86db14d9568953cab4f6a8 down. Removing records
INFO: 2015/07/29 17:22:58.112861 [zonedb] Container 6e1c885eb52cb2bc27e04f8361e22668fbb264f2d0db9dceaf9682d6d57cbd4d down. Removing records
INFO: 2015/07/29 17:24:52.534833 [dns msgid 36255] No results for type A query for 'pinghim.weave.local.' [caching no-local]
WARNING: 2015/07/29 17:24:52.792157 [dns msgid 10231] Failed lookup for external name pinghim.
WARNING: 2015/07/29 17:24:57.887972 [dns msgid 56874] Failed lookup for external name pinghim.
INFO: 2015/07/30 10:10:25.011302 [zonedb] Container 83d2eeddb03f8740c86f091ed21829f6aaec94df307bff58ec9aab99b1dd34b9 down. Removing records
INFO: 2015/07/30 10:10:29.821714 [zonedb] Container 25190080cd5dfc0ece529d6a32de509108093735143e191ff07eacce07c8dd0a down. Removing records
INFO: 2015/07/30 10:10:58.108913 [http] Adding test.weave.local -> 10.128.0.3
INFO: 2015/07/30 10:11:10.824444 [http] Adding test.weave.local -> 10.128.0.4
INFO: 2015/07/30 10:11:30.896999 [dns msgid 24305] No results for type A query for 'pingthem.weave.local.' [caching no-local]
WARNING: 2015/07/30 10:11:32.329307 [dns msgid 5674] Failed lookup for external name pingthem.
[root@tivx94 ~]# 
```

And here are messages <code>ping -R</code> from <code>pinghim</code> container which was created yesterday. 

```
[root@tivx94 ~]# docker exec pinghim ping -R test.weave.local
PING test.weave.local (10.128.0.4) 56(124) bytes of data.
64 bytes from test.weave.local (10.128.0.4): icmp_seq=1 ttl=64 time=0.117 ms
RR:     pingme.weave.local (10.128.0.2)
        test.weave.local (10.128.0.4)
        test.weave.local (10.128.0.4)
        pingme.weave.local (10.128.0.2)

64 bytes from test.weave.local (10.128.0.4): icmp_seq=2 ttl=64 time=0.227 ms    (same route)
64 bytes from test.weave.local (10.128.0.4): icmp_seq=3 ttl=64 time=0.129 ms    (same route)
^C[root@tivx94 ~]# docker exec pinghim cat /etc/resolv.conf
nameserver 10.10.10.1
[root@tivx94 ~]# 
```

And <code>ping -R </code> from <code>pingthem</code> and <code>pingtest</code> which were own the hostname <code> test.weave.local</code>, then it was resolved to docker network IP.

```
[root@tivx94 ~]# docker exec pingthem ping -R test.weave.local
PING test.weave.local (10.10.10.25) 56(124) bytes of data.
64 bytes from test.weave.local (10.10.10.25): icmp_seq=1 ttl=64 time=0.059 ms
RR:     test.weave.local (10.10.10.25)
        test.weave.local (10.10.10.25)
        test.weave.local (10.10.10.25)
        test.weave.local (10.10.10.25)

64 bytes from test.weave.local (10.10.10.25): icmp_seq=2 ttl=64 time=0.078 ms   (same route)
64 bytes from test.weave.local (10.10.10.25): icmp_seq=3 ttl=64 time=0.078 ms   (same route)
64 bytes from test.weave.local (10.10.10.25): icmp_seq=4 ttl=64 time=0.077 ms   (same route)
[root@tivx94 ~]#
```

**Comment by rade** [comment]
> ````
[root@tivx94 ~]# docker run -dti --hostname=test.weave.local --name pingthem  fedora/php-netcat:4.2.8 nc -lk -p 4300
def9924c7e44e478f8e8b6997ce0f1f243dfba462e4809f58408a03cafed42f7
[root@tivx94 ~]# docker run -dti --hostname=test.weave.local --name pingtest  fedora/php-netcat:4.2.8 nc -lk -p 4300
bff31b3211c86db1305e8561db04587aa263eae7ab701313bad66bc5808cb0df
````

Here you are giving both containers the same hostname. As well as a container name. From [our docs](http://docs.weave.works/weave/latest_release/weavedns.html#usage): "if both hostname and container name are specified at the same time the hostname takes precedence".

Usually there is no need to specify the hostname; it gets derived automatically from the container name.

**Comment by rade** [comment]
> `docker exec pingtest ping test.weave.local`

The reason that returns the docker IP is because you set the hostname of the `pingtest` container to `test.weave.local`. Docker adds an entry to the container's /etc/hosts for the container's hostname and docker IP. That takes precedence over a DNS lookup. Fixed in #1079.
